demo/rust/micdemo/src/main.rs

```diff
@@ -72,79 +46,67 @@ fn porcupine_demo(
-    println!("Using {:?} backend", miniaudio_context.backend());
-
-    let mut device_config = miniaudio::DeviceConfig::new(miniaudio::DeviceType::Capture);
-    device_config
-        .capture_mut()
-        .set_format(miniaudio::Format::S16);
-    device_config.capture_mut().set_channels(1);
-    device_config.capture_mut().set_device_id(Some(device_id));
-    device_config.set_sample_rate(porcupine.sample_rate());
-    device_config.set_data_callback(move |_, _, frames| {
-        process_audio_chunk(
-            frames.as_samples(),
-            &porcupine,
-            &mut buffer,
-            &keywords_or_paths,
-        );
-
-        if let Some(output_file_writer_mutex) = &output_file_writer {
-            let mut output_file_writer = output_file_writer_mutex.lock().unwrap();
-            let samples: &[i16] = frames.as_samples();
-            for sample in samples {
-                output_file_writer.write_sample(*sample).unwrap();
-            }
-        }
-    });
-
-    let device = miniaudio::Device::new(Some(miniaudio_context), &device_config)
-        .expect("Failed to initialize capture device");
+    let recorder = RecorderBuilder::new()
+        .device_index(audio_device_index)
+        .frame_length(porcupine.frame_length() as i32)
+        .init()
+        .expect("Failed to initialize pvrecorder");
+    recorder.start().expect("Failed to start audio recording");
 
     LISTENING.store(true, Ordering::SeqCst);
-    device.start().expect("Failed to start device");
-    println!("Listening for wake words...");
-
     ctrlc::set_handler(|| {
         LISTENING.store(false, Ordering::SeqCst);
     })
     .expect("Unable to setup signal handler");
 
+    println!("Listening for wake words...");
+
+    let mut audio_data = Vec::new();
     // Spin loop until we receive the ctrlc handler
     while LISTENING.load(Ordering::SeqCst) {
         std::hint::spin_loop();
         std::thread::sleep(std::time::Duration::from_millis(10));
+        let mut pcm = vec![0; recorder.frame_length()];
+        recorder
+            .read(&mut pcm)
+            .expect("Failed to read audio frame");
+
+        let keyword_index = porcupine.process(&pcm).unwrap();
+        if keyword_index >= 0 {
+            println!(
+                "[{}] Detected {}",
+                Local::now().format("%F %T"),
+                keywords_or_paths.get(keyword_index as usize)
+            );
+        }
+
+        if !output_path.is_none() {
+            audio_data.extend_from_slice(&pcm);
+        }
     }
 
-    println!("\nStopping!");
-    device.stop().expect("Failed to stop device");
-    println!("Stopped");
+    println!("\nStopping...");
+    recorder.stop().expect("Failed to stop audio recording");
+
+    if !output_path.is_none() {
+        let wavspec = hound::WavSpec {
+            channels: 1,
+            sample_rate: porcupine.sample_rate(),
+            bits_per_sample: 16,
+            sample_format: hound::SampleFormat::Int,
+        };
```

Comment:
We can remove the `if` statement and move this block into the `Some(..)` path of the `match` statement. Otherwise we are checking for "someness" twice.

Reply:
Actually I think a more idiomatic way would be to drop the `match` statement entirely since we don't do anything with the the `none` branch, you can use the `if let` syntax for this:

```rust
if let Some(output_path) = output_path {
...
}
```
This is `true` if output_path is `Some` and then gives us the inner value of the output_path `Option` inside the block, and `false` otherwise.